### PR TITLE
Guard unstable proc_macro things behind `nightly` feature

### PR DESCRIPTION
**Pull request by LukasKalbertodt**
This does most of the work for #13. The crate now has a `nightly` feature that is disabled by default. With that feature disabled, we use not-so-nice diagnostics and call-site hygiene. With the feature enabled, we use nice diagnostics and def-site hygiene.

It still doesn't compile on stable, as it uses several Rust 2018 features (as described in #13). But again, those will be stabilized soon. And even if not, they are easy to replace.

The second commit adds two examples that showcase error messages. It's a bit of a hack... I don't know if this is OK like that. (EDIT: I think I don't quite like it. I would probably instead try to add an option to the compile-fail script to just compile a specific example without capturing stdout. But maybe we can still merge it like this and I can change it later.)

### CODE REVIEW

File: src/diag.rs [hunk 1]
```diff
@@ -1,10 +1,26 @@
-use proc_macro::{Diagnostic, Span};
+//! This module has two purposes:
+//!
+//! 1. Provide the convenienve method `emit_with_attr_note` and add it via
```
Comment (KodrAus) [inline src/diag.rs]: nit: little typo here :)

**Comment by LukasKalbertodt** [comment]
> I'll try spend a little more time understanding how we're shimming the stable vs nightly APIs and what our plan to unify them is.

Sure! If you have any questions or discuss anything, you can also just write via email or ping me on Discord (I'm rarely on IRC, sorry!).

But yes, I'm not completely satisfied with these changes. In particular, fewer `#[cfg]` attributes would be nice. But yeah that's the best I came up with and I figured it's fine for now.

---

Regarding the examples that are supposed to show an error message: what do you think about removing this cfg hack and instead just comment-out the line that's causing the error. So something like:

```rust
    // Add the line below to see the error message.
    //#[auto_impl(Boxxi)]
    trait Foo {
        fn foo(&self) -> u32;
    }
```

Then we could also combine the fail examples into one example `error_messages` or so.

I think I would like to change that still. What do you think?

**Comment by LukasKalbertodt** [comment]
@KodrAus I added a bunch of new commits. The commit messages should explain everything, but the main change is to update several files to a newer nightly version (in particular because of https://github.com/rust-lang/rust/pull/50911). Basically, we have fewer `#![feature(...)]` and fewer `extern crate` now.

The last commit simplifies the examples. It does what I suggested in the last comment. I think I'm happier with it this way.

**Comment by KodrAus** [comment]
Yeh this looks nice and clear :+1: Feel free to merge whenever you're ready!

**Comment by LukasKalbertodt** [comment]
Wonderful :)

**Pull request merged**